polkadot/xcm/xcm-executor/src/traits/asset_transfer.rs

```diff
@@ -0,0 +1,84 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use frame_support::traits::{ContainsPair, PalletError};
+use scale_info::TypeInfo;
+use sp_runtime::codec::{Decode, Encode};
+use xcm::prelude::*;
+
+#[derive(Copy, Clone, Encode, Decode, Eq, PartialEq, Debug, TypeInfo)]
+pub enum Error {
+	/// Invalid non-concrete asset.
+	NotConcrete,
+	/// Invalid non-fungible asset.
+	NotFungible,
+	/// Reserve chain could not be determined for assets.
+	UnknownReserve,
+}
+
+impl PalletError for Error {
+	const MAX_ENCODED_SIZE: usize = 1;
+}
+
+#[derive(Copy, Clone, PartialEq, Debug)]
+pub enum TransferType {
+	Teleport,
+	LocalReserve,
+	DestinationReserve,
```

Comment:
we should probably add here some docs for every type

Reply:
done